include/drivers/misc/ft8xx/ft8xx_common.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2020 Hubert Miś
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief FT8XX common functions
+ */
+
+#ifndef ZEPHYR_DRIVERS_MISC_FT8XX_FT8XX_COMMON_H_
+#define ZEPHYR_DRIVERS_MISC_FT8XX_FT8XX_COMMON_H_
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief FT8xx functions to write and read memory
+ * @ingroup ft8xx_interfaces
+ * @{
+ */
+
+/** Short notation to write 1 byte to FT8xx memory. */
+#define wr8 ft8xx_wr8
+/** Short notation to write 2 bytes to FT8xx memory. */
+#define wr16 ft8xx_wr16
+/** Short notation to write 4 bytes to FT8xx memory. */
+#define wr32 ft8xx_wr32
```

Comment:
Please make the whole defines and wrappers away and use the functions directly (also all the encode and parse wrappers).

Reply:
These wrappers are needed to align with API defined in FT800 Programmer Guide:
![image](https://user-images.githubusercontent.com/21953319/111278943-6ffcdd80-863a-11eb-97a3-fb89e7291362.png)

Reply:
How that applies to this Zephyr OS driver or why do you have to follow it? What am I missing?

Reply:
It's like POSIX API for sockets. There are multiple libraries using POSIX API and now you can easily port those libraries to Zephyr, because Zephyr exposes widely used POSIX socket API.

In similar way, you with this PR you can easily port libraries using FT800 to Zephyr applications, because this driver exposes widely used public API defined in FT800 Programmer Guide.

If I removed these macros, I would need to rewrite each library using FT800 I would like to use in Zephyr-based projects.

Reply:
okay, I see no reason to copy this very bad example. This "API" must have a namespace. Even if this "API"  would have proper namespace prefix it should not be done using macros but inline functions.

> If I removed these macros, I would need to rewrite each library using FT800 I would like to use in Zephyr-based projects.

Well it is not that much work to grep and prefix all the calls with "ft8xx_" or  you could just add a wrapper in your application code.

Reply:
I can easily change macros to inline functions, but I would like to keep these in Zephyr driver. Moving it to each library integration code is more complicated for me, than maintaining my own FT800 driver for Zephyr in project repositories.

I see no point in sharing my driver with community if I cannot easily use it in my projects.

Reply:
I think that if there are a lot of existing FT800 examples using this "API" we should just provide it to ease portability. However, because it is easy to have name clashes with such short names, we should probably restrict its usage to external samples and not code in Zephyr.

Reply:
> it is easy to have name clashes with such short names

That's exactly why I designed Zephyr driver in a way that public symbols have `ft8xx_` prefix, to avoid name collisions. Only compilation units including `ft8xx_common.h`, `ft8xx_copro.h`, or `ft8xx_memory.h` can be subjected to such collisions. Such compilation units are usually responsible for display, touch processing, or audio and do not include other libraries that may be prone to clashes.

Perhaps I could create separated header like `ft8xx_reference_api.h` including only `static inline` mappings of `ft8xx_...()` functions to shorter names. This way we would avoid name conflicts, because only compilation units including this header would be troubled. And there would be no other reason to include such file than to use this defined API.

What do you think?

Reply:
The problem may appear if Zephyr ever exposes such names via a sufficiently common header. I think the best we can do is to restrict such helpers to external applications (maybe sample at most for now).

Reply:
> Moving it to each library integration code is more complicated for me, than maintaining my own FT800 driver for Zephyr in project repositories.
> 
> I see no point in sharing my driver with community if I cannot easily use it in my projects.

I do not see how this PR can go in if the public driver specific "API" does not use a namespace, regardless via what public header it is exposed. Others can argue just like you, why should they not get the right to use the same identifiers? How should this drivers work together in an application?

Reply:
> How should this drivers work together in an application?

Obviously if you use multiple poorly designed APIs (like `wr32()`) inside single compilation unit, it cannot work correctly. However, this driver does not expose any global symbols without "namespace" prefixes. These short names are affecting only compilation units that include header files from this driver.

I agree with @gmarull that we should limit these includes in Zephyr to only samples to make sure that these files are not included indirectly.

I'll prepare a proposal in the evening separating these non-namespace from the rest of the code. I hope such solution will be acceptable by everyone. We can discuss later after I push it.

Reply:
I've moved not-prefixed names from `ft8xx_common.h` and `ft8xx_copro.h` to a new header file `ft8xx_reference_api.h`. This new header file is included only in the FT800 sample (and in platform integration in my projects). This way only single sample main.c compilation unit is polluted with this potentially conflicting API. Problematic header cannot be unintentionally indirectly included, because none of headers include it. I've also added a note to the documentation of this header to use it with caution because of potential naming conflicts.

Reply:
This sample is also what I am talking about, something like `cmd(CLEAR_COLOR_RGB(0x00, 0x00, 0x00));` implies `cmd()` is some common API but it is not. This sample should only use the prefixed driver's specific API, ft8xxx_cmd() and not cmd(). People will use it and take this bad example for their own application or even worse for another driver.

Reply:
I agree that we should show using `ft8xx_` prefixed names as an example. I'm only concerned that no code in Zephyr includes `ft8xx_reference_api.h` now and because of this it's not easy to even verify if syntax in this file is correct using a compiler.

However, this file is relatively simple and I checked it with an out-of-tree application. It's not a big problem to maintain it without using in Zephyr directly.